ui_test/src/rustc_stderr.rs

```diff
@@ -0,0 +1,144 @@
+use std::{
+    fmt::Write,
+    path::{Path, PathBuf},
+};
+
+use regex::Regex;
+
+#[derive(serde::Deserialize, Debug)]
+struct RustcMessage {
+    rendered: Option<String>,
+    spans: Vec<Span>,
+    level: String,
+    message: String,
+    children: Vec<RustcMessage>,
+}
+
+#[derive(Copy, Clone, Debug, PartialOrd, Ord, PartialEq, Eq)]
+pub(crate) enum Level {
+    Error = 0,
+    Warn = 1,
+    Help = 2,
+    Note = 3,
+    /// Only used for "For more information about this error, try `rustc --explain EXXXX`".
+    FailureNote = 4,
+}
+
+impl std::str::FromStr for Level {
+    type Err = String;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "ERROR" | "error" => Ok(Self::Error),
+            "WARN" | "warning" => Ok(Self::Warn),
+            "HELP" | "help" => Ok(Self::Help),
+            "NOTE" | "note" => Ok(Self::Note),
+            "failure-note" => Ok(Self::FailureNote),
+            _ => Err(format!("unknown level `{s}`")),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct Message {
+    pub(crate) level: Level,
+    pub(crate) message: String,
+}
+
+impl RustcMessage {
+    fn line(&self, file: &Path) -> Option<usize> {
+        self.spans.iter().find_map(|span| span.line(file))
+    }
+
+    /// Put the message and its children into the line-indexed list.
+    fn insert_recursive(
+        self,
+        file: &Path,
+        messages: &mut Vec<Vec<Message>>,
+        messages_from_unknown_file_or_line: &mut Vec<Message>,
+        line: Option<usize>,
+    ) {
+        let line = self.line(file).or(line);
```

Comment:
The `line` argument is equal to `self.line(file)` (as passed from `process`)... so I don't understand what this logic is good for. Is it something with recursive diagnostics?

Reply:
Yea, this is just for the recursive ones. I guess I can just pass in `None` at the root invocation

Reply:
Okay, then please also explain the point of this parameter in a doc comment.